clang/lib/AST/Interp/ByteCodeExprGen.h

```diff
@@ -478,6 +480,43 @@ template <class Emitter> class SourceLocScope final {
   bool Enabled = false;
 };
 
+template <class Emitter> class StoredOpaqueValueScope final {
+public:
+  StoredOpaqueValueScope(ByteCodeExprGen<Emitter> *Ctx) : Ctx(Ctx) {}
+
+  bool VisitAndStoreOpaqueValue(const OpaqueValueExpr *Ove) {
+    assert(Ove && "OpaqueValueExpr is a nullptr!");
+    assert(!Ctx->OpaqueExprs.contains(Ove) &&
+           "OpaqueValueExpr already stored!");
+
+    std::optional<PrimType> CommonTy = Ctx->classify(Ove);
+    std::optional<unsigned> LocalIndex = Ctx->allocateLocalPrimitive(
+        Ove, *CommonTy, Ove->getType().isConstQualified());
+    if (!LocalIndex)
+      return false;
+
+    if (!Ctx->visit(Ove))
+      return false;
+
+    if (!Ctx->emitSetLocal(*CommonTy, *LocalIndex, Ove))
+      return false;
+
+    Ctx->OpaqueExprs.insert({Ove, *LocalIndex});
+    StoredValues.emplace_back(Ove);
+
+    return true;
```

Comment:
I think all the visiting logic should go into `ByteCodeExprGen.cpp`.

Reply:
I wanted to follow the same pattern that other scopes in the same header file have, when the functions are defined inside the header, but I'll move this definition into the source file as requested.

Reply:
Sorry, I meant that the visiting of the nodes and the creation of the local variable should move to `VisitArrayInitLoopExpr` instead. But now that I look at it -- shouldn't this be in `VisitOpaqueValueExpr`?

Reply:
I moved the "caching" logic into `VisitOpaqueValueExpr`.